python/gps/algorithm/algorithm_traj_opt.py

```diff
@@ -98,15 +105,107 @@ def _stepadjust(self, m):
+            if self._hyperparams['demo_distr_empest']:
+                demo_traj[i] = fit_emp_controllers(demoX[i], demoU[i])
+            # Use LQR estimations?
+
+        for i in xrange(M):
+            # This code assumes a fixed number of samples per iteration/controller
+            samples_prob[i] = np.zeros((itr + Md, self.T, (self.N / M) * (itr - 1) + init_samples))
+            demos_prob[i] = np.zeros((itr + Md, self.T, demoX[i].shape[0]))
+            sample_i_X = self.cur[i].sample_list.get_X()
+            sample_i_U = self.cur[i].sample_list.get_U()
+            # Evaluate sample prob under sample distributions
+            for itr_i in xrange(itr):
+                traj = self.traj_distr[itr_i][i]
+                for j in xrange(sample_i_X.shape[0]):
+                    for t in xrange(self.T - 1):
+                        diff = traj.k[t, :] + \
+                                traj.K[t, :, :].dot(sample_i_X[j, t, :]) - sample_i_U[j, t, :]
+                        samples_prob[i][itr_i, t, j] = -0.5 * np.sum(diff * (traj.inv_pol_covar[t, :, :].dot(diff)), 0) - \
+                                                        np.sum(np.log(np.diag(traj.chol_pol_covar[t, :, :])))
+
+            # Evaluate sample prob under demo distribution.
+            for itr_i in xrange(Md):
+                for j in range(sample_i_X.shape[0]):
+                    for t in xrange(self.T - 1):
+                        diff = demo_traj[itr_i].k[t, :] + \
+                                demo_traj[itr_i].K[t, :, :].dot(sample_i_X[j, t, :]) - sample_i_U[j, t, :]
+                        samples_prob[i][itr + itr_i, t, j] = -0.5 * np.sum(diff * (demo_traj[itr_i].inv_pol_covar[t, :, :].dot(diff)), 0) - \
+                                                        np.sum(np.log(np.diag(demo_traj[itr_i].chol_pol_covar[t, :, :])))
+            # Sum over time.
+            samples_prob_T[i] = np.sum(samples_prob[i], 1)
+            # Sum over the distributions.
+            samples_sum[i] = logsum(samples_prob_T[i], 0)
+
+        # Assume only one condition for the samples.
+        assert Md == M or M == 1
+        for idx in xrange(Md):
+            if M == 1:
+                i = 1
+            else:
+                i = idx
+            # Evaluate demo prob. under sample distributions.
+            for itr_i in xrange(itr):
+                traj = self.traj_distr[itr_i][i]
+                for j in xrange(demoX[idx].shape[0]):
+                    for t in xrange(self.T - 1):
+                        diff = traj.k[t, :] + \
+                                traj.K[t, :, :].dot(demoX[idx][j, t, :]) - demoU[idx][j, t, :]
+                        demos_prob[idx][itr_i, t, j] = -0.5 * np.sum(diff * (traj.inv_pol_covar[t, :, :].dot(diff)), 0) - \
+                                                        np.sum(np.log(np.diag(traj.chol_pol_covar[t, :, :])))
+            # Evaluate demo prob. under demo distributions.
+            for itr_i in xrange(Md):
```

Comment:
I think that we can assume that Md is always equal to 1. I was experimenting before with fitting more than one empirical distribution to the demos, hence this code in the matlab codebase, but I think we should just have Md=1 moving forward.